packages/firestore/src/local/instance_metadata_channel.ts

```diff
@@ -0,0 +1,455 @@
+      activeTargetIdsSet = activeTargetIdsSet.add(activeTargetIdsArray[i]);
+    }
+
+    if (validData) {
+      return new InstanceState(
+        instanceKey,
+        new Date(instanceData.lastUpdateTime),
+        activeTargetIdsSet,
+        instanceData.minMutationBatchId,
+        instanceData.maxMutationBatchId
+      );
+    }
+
+    error(
+      LOG_TAG,
+      `Failed to parse instance metadata for instance '${instanceKey}'`
+    );
+    return null;
+  }
+}
+
+/**
+ * Instance state of the local instance. Unlike `InstanceState`, this class
+ * is mutable and keeps track of all pending mutations, which allows us to
+ * update the range of pending mutation batch IDs as new mutations are added or
+ * removed.
+ *
+ * The data in `LocalInstanceState` is not read from LocalStorage and instead
+ * updated via its instance methods. The updated state can be serialized via
+ * `toLocalStorageJSON()`.
+ */
+class LocalInstanceState extends InstanceState {
+  private pendingBatchIds = new SortedSet<BatchId>(primitiveComparator);
+
+  constructor(instanceKey: InstanceKey) {
+    super(
+      instanceKey,
+      new Date(),
+      new SortedSet<TargetId>(primitiveComparator),
+      null,
+      null
+    );
+  }
+
+  get minMutationBatchId(): BatchId | null {
+    return this.pendingBatchIds.first();
+  }
+
+  get maxMutationBatchId(): BatchId | null {
+    return this.pendingBatchIds.last();
+  }
+
+  addPendingMutation(batchId: BatchId): void {
+    assert(
+      !this.pendingBatchIds.has(batchId),
+      `Batch with ID '${batchId}' already pending.`
+    );
+    this.pendingBatchIds = this.pendingBatchIds.add(batchId);
+  }
+
+  removePendingMutation(batchId: BatchId): void {
+    assert(
+      this.pendingBatchIds.has(batchId),
+      `Pending Batch ID '${batchId}' not found.`
+    );
+    this.pendingBatchIds = this.pendingBatchIds.delete(batchId);
+  }
+
+  addQueryTarget(targetId: TargetId): void {
+    assert(
+      !this.activeTargetIds.has(targetId),
+      `Target with ID '${targetId}' already active.`
+    );
+    this.activeTargetIds = this.activeTargetIds.add(targetId);
+  }
+
+  removeQueryTarget(targetId: TargetId): void {
+    assert(
+      this.activeTargetIds.has(targetId),
+      `Active Target ID '${targetId}' not found.`
+    );
+    this.activeTargetIds = this.activeTargetIds.delete(targetId);
+  }
+
+  /** Sets the update time to the current time. */
+  refreshLastUpdateTime(): void {
+    this.lastUpdateTime = new Date();
+  }
+
+  /**
+   * Converts this entry into a JSON-encoded format we can use for LocalStorage.
+   * Does not encode `instanceKey` as it is part of the key in LocalStorage.
+   */
+  toLocalStorageJSON(): string {
+    const data: InstanceStateSchema = {
+      lastUpdateTime: this.lastUpdateTime.getTime(),
+      activeTargetIds: []
+    };
+
+    this.activeTargetIds.forEach(targetId => {
+      data.activeTargetIds.push(targetId);
+    });
```

Comment:
If you feel like it, I'd be a fan of adding SortedSet.toArray() method to simplify this code.